cores/nRF5/common_func.h

```diff
@@ -37,17 +37,48 @@
 #ifndef _COMMON_FUNC_H_
 #define _COMMON_FUNC_H_
 
+
 #ifdef __cplusplus
- extern "C" {
+  // namespace and templates must be outside the `extern "C"` declaration...
+  namespace ADAFRUIT_DETAIL
```

Comment:
Thank you for you patient, I should have done this long ago. Since I find this very interesting and like to do a bit of research myself, but couldn't find time. So yeah, it is not really the reason of delay but worth mentionning :sweat_smile: 

I like the C-only version more than the C++ version, can be just use the C-only version for this. If not then I am happy to merge it.

Reply:
Hi @hathach,

Thanks for the review.  

It has been some time since I wrote this.  I have tested only the C++ version when compiling C++ source files, and have tested only the C version when compiling C source files.  Also, the C version requires GCC extensions, while the C++ version is standards-compliant.

One large benefit of separate C vs C++ versions is that the C++ version results in more clear error messages.  After all, an error message of `"error: negative width in bit-field '<anonymous>'"` is not really clear as to the cause.  In contrast, an error message that includes the function name `arrcount_fails_if_not_array`, is substantially more clear as to the problem.

Of course, if you are happy with the C version error message, then please just confirm this is your preference, and I will adjust lines 41 and 54 accordingly.

then simply exclude the same as I think the C version is less capable than the C++ version ... for example, it requires GCC extensions, which perhaps not all compilers support.

However, please feel free to modify this branch in any way you see fit.  If the C version works in all cases, and you are happy with the

Reply:
@henrygab thanks for clarification, I merge it now.